File: frontend/src/api/records.ts
import { http } from "./http";
import type { PracticeProjectDetailResponse, PracticeProjectListResponse, RecordFlagsResponse, RecordWorkspaceResponse } from "../types/video";

export async function getRecordsWorkspace(limit = 50): Promise<RecordWorkspaceResponse> {
  const { data } = await http.get<RecordWorkspaceResponse>("/api/records/workspace", {
    params: { limit },
  });
  return data;
}

export async function getRecordFlags(): Promise<Record<string, { starred?: boolean; note?: string; updated_at?: string }>> {
  const { data } = await http.get<{ items: Record<string, { starred?: boolean; note?: string; updated_at?: string }> }>("/api/records/flags");
  return data.items || {};
}

export async function updateRecordFlags(pipelineId: string, payload: { starred?: boolean; note?: string }): Promise<RecordFlagsResponse> {
  const { data } = await http.patch<RecordFlagsResponse>(`/api/records/${encodeURIComponent(pipelineId)}/flags`, payload);
  return data;
}

export async function downloadRecordsCsv(params: {
  limit?: number;
  starred?: boolean;
  status?: "done" | "failed" | "running";
} = {}): Promise<Blob> {
  const { data } = await http.get<Blob>("/api/records/export.csv", {
    params,
    responseType: "blob",
  });
  return data;
}

export async function getPracticeProjects(limit = 200): Promise<PracticeProjectListResponse> {
  const { data } = await http.get<PracticeProjectListResponse>("/api/practice-projects", {
    params: { limit },
  });
  return data;
}

export async function getPracticeProject(teacherVideoId: string, limit = 200): Promise<PracticeProjectDetailResponse> {
  const { data } = await http.get<PracticeProjectDetailResponse>(`/api/practice-projects/${encodeURIComponent(teacherVideoId)}`, {
    params: { limit },
  });
  return data;
}

File: backend/app/api/records.py
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Response

from app.schemas.dto import RecordFlagsPatchRequest, RecordFlagsResponse, RecordWorkspaceResponse
from app.services.record_export import build_records_csv
from app.services.record_flags import load_record_flags, update_record_flags
from app.services.records import list_records_workspace

router = APIRouter(prefix="/records", tags=["records"])


@router.get("/workspace", response_model=RecordWorkspaceResponse)
async def records_workspace_api(limit: int = 50):
    return RecordWorkspaceResponse(**list_records_workspace(limit=limit))


@router.get("/flags")
async def record_flags_api():
    return {"items": load_record_flags()}


@router.get("/export.csv")
async def export_records_csv_api(limit: int = 200, starred: bool | None = None, status: str | None = None):
    allowed_status = {"done", "failed", "running"}
    normalized_status = str(status).strip() if status is not None else None
    if normalized_status not in allowed_status:
        normalized_status = None
    content = build_records_csv(limit=limit, starred=starred, status=normalized_status)
    filename = f"dance_assist_records_{date.today().isoformat()}.csv"
    return Response(
        content=content,
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.patch("/{pipeline_id}/flags", response_model=RecordFlagsResponse)
async def patch_record_flags_api(pipeline_id: str, payload: RecordFlagsPatchRequest):
    return RecordFlagsResponse(**update_record_flags(pipeline_id, starred=payload.starred, note=payload.note))
